fem/quadinterpolator.hpp

```diff
@@ -0,0 +1,222 @@
+                          const IntegrationRule &ir);
+
+   QuadratureInterpolator(const FiniteElementSpace &fes,
+                          const QuadratureSpace &qs);
+
+   /** @brief Disable the use of tensor product evaluations, for tensor-product
+       elements, e.g. quads and hexes. */
+   /** Currently, tensor product evaluations are not implemented and this method
+       has no effect. */
+   void DisableTensorProducts(bool disable = true) const
+   { use_tensor_products = !disable; }
+
+   /// Interpolate the E-vector @a e_vec to quadrature points.
+   /** The @a eval_flags are a bitwise mask of constants from the EvalFlags
+       enumeration. When the VALUES flag is set, the values at quadrature points
+       are computed and stored in the Vector @a q_val. Similarly, when the flag
+       DERIVATIVES is set, the derivatives are computed and stored in @a q_der.
+       When the DETERMINANTS flags is set, it is assumed that the derivatives
+       form a matrix at each quadrature point (i.e. the associated
+       FiniteElementSpace is a vector space) and their determinants are computed
+       and stored in @a q_det. */
+   void Mult(const Vector &e_vec, unsigned eval_flags,
+             Vector &q_val, Vector &q_der, Vector &q_det) const;
+
+   /// Perform the transpose operation of Mult(). (TODO)
+   void MultTranspose(unsigned eval_flags, const Vector &q_val,
+                      const Vector &q_der, Vector &e_vec) const;
+
+   // Compute kernels follow (cannot be private or protected with nvcc)
+
+   /// Template compute kernel for 2D.
+   template<const int T_VDIM = 0, const int T_ND = 0, const int T_NQ = 0>
+   static void Eval2D(const int NE,
+                      const int vdim,
+                      const DofToQuad &maps,
+                      const Vector &e_vec,
+                      Vector &q_val,
+                      Vector &q_der,
+                      Vector &q_det,
+                      const int eval_flags);
+
+   /// Template compute kernel for 3D.
+   template<const int T_VDIM = 0, const int T_ND = 0, const int T_NQ = 0>
+   static void Eval3D(const int NE,
+                      const int vdim,
+                      const DofToQuad &maps,
+                      const Vector &e_vec,
+                      Vector &q_val,
+                      Vector &q_der,
+                      Vector &q_det,
+                      const int eval_flags);
+};
+
+/** @brief A class that performs interpolation from an E-vector to quadrature
+    point values and/or derivatives (Q-vectors) on the faces. */
+/** An E-vector represents the element-wise discontinuous version of the FE
+    space and can be obtained, for example, from a GridFunction using the
+    Operator returned by FiniteElementSpace::GetElementRestriction().
+
+    The target quadrature points in the elements can be described either by an
+    IntegrationRule (all mesh elements must be of the same type in this case) or
+    by a QuadratureSpace. */
```

Comment:
Similarly, "elements" -> "faces"; also, the `QuadratureSpace` is not quite suitable for describing face quadratures, so we should probably remove that part.